crates/store/src/db/migrations.rs

```diff
@@ -114,16 +114,46 @@ fn compute_migration_hashes() -> Vec<Hash> {
 }
 
 fn preprocess_sql(sql: &str) -> String {
-    // TODO: We can also remove all comments here (need to analyze the SQL script in order to remove
-    //       comments in string literals).
-    remove_spaces(sql)
+    remove_unnecessary(sql)
 }
 
-fn remove_spaces(str: &str) -> String {
-    str.chars().filter(|chr| !chr.is_whitespace()).collect()
+// Strips new lines, whitespaces and comments.
+fn remove_unnecessary(str: &str) -> String {
+    let mut rslt = String::with_capacity(32);
+    for line in str.lines() {
+        let mut iter = line.as_bytes().iter().copied();
+        let Some(mut prev) = iter.next() else {
+            continue;
+        };
+        rslt.push(prev.into());
+        for curr in iter {
+            if prev == b'-' && curr == b'-' {
+                rslt.pop();
+                break;
+            }
+            if !curr.is_ascii_whitespace() {
+                rslt.push(curr.into());
+            }
+            prev = curr;
+        }
+    }
+    rslt
```

Comment:
This works but I think we can make this more legible by using iterators:

```suggestion
/// Processes a sql migration to ensure that non-syntactic diffs don't change the hash.
///
/// This is done by removing all comments and whitespace.
fn preprocess_migration_script(sql: &str) -> String {
    sql
      .lines()
      .map(|line| line.split_once("--").unwrap_or((line, "")).0)
      .flat_map(|line| line.chars())
      .filter(|c| !c.is_whitespace())
      .collect()
}
```

I would also replace `preprocess_sql` with this.

Reply:
In Rust iterators are historically known for generally not yielding optimized code. For example:

- https://github.com/rust-lang/rust/issues/47461
- https://github.com/rust-lang/rust/issues/117659
- https://github.com/rust-lang/rust/issues/113691
- https://github.com/rust-lang/rust/issues/75935

[This godbolt snippet](https://godbolt.org/z/Gqa4PjTve) also states that the `Iterator` version produces way more lines of assembly.

In regards to runtime performance, the `Iterator` version is 2x slower than the `loop` version.

![Capture d’écran du 2025-06-27 08-23-23](https://github.com/user-attachments/assets/867d40b2-2ad5-4cfc-a5c3-40c8aa88300f)

```rust
#[bench]
fn bench_iter(b: &mut Bencher) {
  let string = String::from_utf8(vec![0; 1024 * 1024 * 2]).unwrap();
  b.iter(|| {
    let _processed = preprocess_iter(&string);
  });
}

#[bench]
fn bench_loop(b: &mut Bencher) {
  let string = String::from_utf8(vec![0; 1024 * 1024 * 2]).unwrap();
  b.iter(|| {
    let _processed = preprocess_loop(&string);
  });
}

fn preprocess_iter(sql: &str) -> String {
  sql
    .lines()
    .map(|line| line.split_once("--").unwrap_or((line, "")).0)
    .flat_map(|line| line.chars())
    .filter(|c| !c.is_whitespace())
    .collect()
}

fn preprocess_loop(sql: &str) -> String {
  let mut rslt = String::with_capacity(32);
  for line in sql.lines() {
    let mut iter = line.as_bytes().iter().copied();
    let Some(mut prev) = iter.next() else {
      continue;
    };
    rslt.push(prev.into());
    for curr in iter {
      if prev == b'-' && curr == b'-' {
        rslt.pop();
        break;
      }
      if !curr.is_ascii_whitespace() {
        rslt.push(curr.into());
      }
      prev = curr;
    }
  }
  rslt
}
```

Reply:
I understand thanks; but I would prefer something that is easy to read.

The code isn't performance sensitive so it doesn't need to be optimized for speed, but rather for legibility and ease of maintenance.

Reply:
The code has been updated to use the `Iterator` version